### PR TITLE
Add disjoint for Map and IntMap

### LINKED ISSUE #614: Improve isSubsetOf
Currently,  `isSubsetOf s1 s2` checks that `size s1 <= size s2`, and if so invokes `isSubsetOfX` which recursively splits `s2` along the structure of `s1` without further size checks. I doubt this is the right approach.

### Wrong side

Splitting a set takes logarithmic time and allocation, so I believe we should split the *smaller* set along the structure of the *larger* one whenever the sizes are different. This is most easily accomplished by splitting `s1` along the structure of `s2` instead. That changes the shape of the algorithm somewhat, but I think that should be okay.

### More size tests are probably better

A size check takes constant time and no allocation, and can save a split and potentially a lot of comparisons. So I think we should perform recursive size checks. Once we split `s1` at the root of `s2`, I think we should check that the left subtree of `s2` is no smaller than the left side of the split and that the right subtree of `s2` is no smaller than the right side of the split before proceeding recursively. For example, suppose we have `{1,2,3,4,5,6,7}` with a root of `4` and want to check if `{0,1,2,3,4,5,6}` is a subset. After the first split, we'll immediately see that it's not, because `{0,1,2,3}` has more elements than `{1,2,3}`.

### LINKED ISSUE #641: Checking IntMaps for disjointness
Is this code optimal?

```
disjoint :: IntMap a -> IntMap b -> Bool
disjoint a b = null (intersection a b)
```

At least I think that it's non-obvious that it's optimal, so it would be good to add it to `IntMap` (and analogously `Map`).

### PR DESCRIPTION
**Pull request by sjakobi**
Closes #641.

### CODE REVIEW

**Comment by sjakobi** [comment]
The tricky parts are all copied from the respective set versions.

**Comment by sjakobi** [comment]
> The tricky parts are all copied from the respective set versions.

This includes the complexity annotations! :)

I probably wouldn't get them right on my own, so please check!

**Comment by treeowl** [comment]
As I recall, `Data.Set.disjoint` has some explanation of what it's doing and why it has the stated complexity. Please find those notes and leave a comment pointing to them from your `Data.Map` version.

**Comment by sjakobi** [comment]
> As I recall, `Data.Set.disjoint` has some explanation of what it's doing and why it has the stated complexity.

I can only find a reference to `subsetOfX`:

https://github.com/haskell/containers/blob/ad3b27ba999b8cb3a00c8c4c6a7da81e3de54189/containers/src/Data/Set/Internal.hs#L689-L693

https://github.com/haskell/containers/blob/ad3b27ba999b8cb3a00c8c4c6a7da81e3de54189/containers/src/Data/Set/Internal.hs#L631-L662

Does any of this apply here?

**Comment by treeowl** [comment]
I suspect it does, but I'm a bit sleepy right now. Maybe you can check?

**Comment by treeowl** [comment]
Oh, I guess just the singleton special cases. Might or might not need an explanation of why we have those. I'll leave it up to you.

**Comment by treeowl** [comment]
Please remove the garbage binary you added, and maybe add a gitignore line to avoid that in the future.

**Comment by sjakobi** [comment]
Oops! I updated my global gitignore.

**Comment by treeowl** [comment]
Please also add a project gitignore exclusion for `*.swo`. We're unlikely to want to commit such a thing, and I'm a bit paranoid about such.

File: containers/src/Data/IntMap/Internal.hs [hunk 1]
```diff
@@ -728,6 +729,42 @@ unsafeFindMax Nil = Nothing
...
+disjoint :: IntMap a -> IntMap b -> Bool
+disjoint Nil _ = True
+disjoint _ Nil = True
+disjoint (Tip kx _) (Tip ky _) = kx /= ky
+disjoint (Tip k _) (Bin p m l r) | nomatch k p m = True
+                                 | zero k m      = notMember k l
+                                 | otherwise     = notMember k r
+disjoint (Bin p m l r) (Tip k _) | nomatch k p m = True
+                                 | zero k m      = notMember k l
+                                 | otherwise     = notMember k r
```
Comment (treeowl) [inline containers/src/Data/IntMap/Internal.hs]: I think you can collapse all the `Tip` cases into two.

```haskell
disjoint (Tip kx _) ys = notMember kx ys
disjoint xs (Tip ky _) = notMember ky xs
```

Comment (sjakobi) [reply containers/src/Data/IntMap/Internal.hs]: :+1:

I've also changed the order of the alternatives so the more likely ones come first.

**Comment by treeowl** [comment]
The order of constructors matched in a `case` expression is utterly irrelevant. GHC reorders them to match the order of constructors in the `data` declaration when lowering from Haskell to Core. Please use the order that's most readable, which usually means shortest first.

**Comment by treeowl** [comment]
Oh, and be sure that all the different `disjoint` things behave the same for `disjoint Nil undefined`.

**Comment by olligobber** [comment]
It looks like it would be easy to add a `disjointWith :: Ord k => (a -> b -> Bool) -> Map k a -> Map k b -> Bool` method, similar to `unionWith`, `isSubmapOfBy`, and `intersectWith`. Should this be added before this is merged?

**Comment by sjakobi** [comment]
> Oh, and be sure that all the different `disjoint` things behave the same for `disjoint Nil undefined`.

```
> Data.IntMap.Internal.disjoint Nil undefined
True
> Data.Map.Internal.disjoint Data.Map.Internal.Tip undefined
True

> Data.Set.Internal.disjoint mempty undefined
True
> Data.IntSet.Internal.disjoint mempty undefined
True
```

**Comment by sjakobi** [comment]
> It looks like it would be easy to add a disjointBy

I personally don't need that right now, and I'm not keen on implementing it at this point. You could add it in a follow-up PR if you want?!

**Comment by sjakobi** [comment]
> The order of constructors matched in a `case` expression is utterly irrelevant. GHC reorders them to match the order of constructors in the `data` declaration when lowering from Haskell to Core. Please use the order that's most readable, which usually means shortest first.

@treeowl Thanks for reminding me! :+1: I'm not sure how I could got that misconception! ;)

**Comment by sjakobi** [comment]
@treeowl Would you mind giving this PR another review?

**Comment by treeowl** [comment]
I expect to release tonight.

**Pull request merged**